client/boot/index.js

```diff
@@ -161,8 +160,39 @@ function boot() {
 
 	translatorJumpstart.init();
 
-	reduxStore = createReduxStore();
+	localforage.config( {
+		name: 'calypso',
+		storeName: 'calypso_store',
+		description: 'Calypso Storage'
+	} );
+	localforage.getItem( 'redux-state', ( error, initialState ) => {
+		let reduxStore;
+		if ( error ) {
+			debug( 'failed to load initial redux-store state', error );
+			reduxStore = createReduxStore();
+		} else {
+			debug( 'loading redux-store with initial state', initialState );
+			reduxStore = createReduxStore( initialState || {} );
+		}
+
+		reduxStore.subscribe( function() {
+			let state = Object.assign( {}, reduxStore.getState() );
+			//remove dirty parts of the redux tree
+			delete state.sites;
+			delete state.themes;
+			localforage.setItem( 'redux-state', state, ( setError ) => {
+				if ( setError ) {
+					debug( 'failed to set redux-store state', setError );
+				}
+			} );
+		} );
+
+		reduxStoreReady( reduxStore );
+	} );
```

Comment:
Definitely. I was also thinking that each module should also verify the initial state that gets passed to it, and add related tests in each module.